ibis/impala/client.py

```diff
@@ -189,7 +189,11 @@ def __init__(self, cursor, con, impyla_con, database,
         self.con.connection_pool_size += 1
 
     def __del__(self):
-        self._close_cursor()
+        try:
+            self._close_cursor()
```

Comment:
Kind of a hack, to prevent `__del__` from raising an exception. Ultimately we should not close cursors for the user. We should depend on context managers here instead.

Reply:
My main concern here is that it's extremely difficult to predict when this method will run because of the reference cycle created when this class is instantiated (`ImpalaCursor` holds a ref to `con` which holds a list of references to `ImpalaCursor`s).

Coupled with this statement in the [`__del__` documentation](https://docs.python.org/3/reference/datamodel.html#object.__del__): `It is not guaranteed that __del__() methods are called for objects that still exist when the interpreter exits.` I don't think defining `__del__` is much better than not defining it.

However, having a way to explicitly close cursors/connections without having to use the `try`/`finally` pattern would at least allow for predictable control of resources.